7 use pydantic for data models

LucasZheng42 opened the pull request:
main changes : 
feat: Integration of Pydantic to use data models
test : update testing data to ensure data models format and merge extract test with transform one and add pydantic model tests
chore(tests): test_load_to_oscar.py can now delete its created files to avoid polluting ECONVENTION_TO_OSCAR_OUTPUT_DIR

LucasZheng42 committed:
feat: Integration of Pydantic to use data models
chore: Refactor to factory design pattern and rename "transform_from_econvention_to_oscar.py" to "transform.py"
test/doc: update testing data to ensure data models format

LucasZheng42 committed:
chore: simplify field_validator by removing duplicate and drop Optional for clarity

LucasZheng42 committed:
fix: update str default value by replacing None by ""

LucasZheng42 committed:
chore(tests): reorganize by splitting one test per file in tasks.

LucasZheng42 committed:
chore(tests): static fixture become global variable and add default values for data_interval variables.
tests: add load test to ensure that data can be loaded as a file before importation.

LucasZheng42 committed:
chore: renaming variables for clarity

LucasZheng42 committed:
chore: remove design pattern factory to simplify the codebase and refactor the project
test : adapt the changes above and merge transform test with extract one.

jdp1ps commented on dags/models/econvention.py:
```diff
@@ -0,0 +1,24 @@
+from pydantic import BaseModel, Field, ConfigDict
+
+
+class Econvention(BaseModel):
```
Rename to "Convention"

LucasZheng42 replied:
done !

jdp1ps commented on dags/models/oscar_convention.py:
```diff
@@ -0,0 +1,44 @@
+from pydantic import BaseModel, Field, PositiveInt, field_validator
+
+
+class OscarConvention(BaseModel):
```
Rename to "Activity"

LucasZheng42 replied:
done !

jdp1ps commented on dags/models/econvention.py:
```diff
@@ -0,0 +1,24 @@
+    SousType: str = ""
+    DateDemarrage: str = ""
+    TermeConvention: str = ""
+    Etape: list[str] = []
```
Use enums or regex validation whenever possible

LucasZheng42 replied:
Used regex for date and enums for somes attributes like "financialImpact"

jdp1ps commented on dags/tasks/econvention_to_oscar/extract.py:
```diff
@@ -0,0 +1,40 @@
+    :return: list of econvention instances formatted into JSON string.
+    """
+    conf_json = context["dag_run"].conf
+    items = conf_json.get("items")
```
use meaningful names : raw_conventions

LucasZheng42 replied:
done !

jdp1ps commented on dags/tasks/econvention_to_oscar/extract.py:
```diff
@@ -0,0 +1,40 @@
+    if not all(isinstance(item, dict) for item in items):
+        raise ValueError("All items in 'items' must be dictionaries.")
+
+    econvention_list: list[Econvention] = []
```
convention_list

LucasZheng42 replied:
done !

jdp1ps commented on dags/tasks/econvention_to_oscar/extract.py:
```diff
@@ -0,0 +1,40 @@
+
+        filtered_data = {
+            k: v for k, v in econvention.items() if k in Econvention.model_fields.keys()
+        }
```
Could be done in Pydantic models. https://docs.pydantic.dev/latest/concepts/alias/

LucasZheng42 replied:
all done with field_validator

jdp1ps commented on dags/tasks/econvention_to_oscar/extract.py:
```diff
@@ -0,0 +1,40 @@
+        filtered_data = {
+            k: v for k, v in econvention.items() if k in Econvention.model_fields.keys()
+        }
+        econvention_list.append(Econvention(**filtered_data))
```
use model_validate

LucasZheng42 replied:
done !

jdp1ps reviewed (commented):
- I can't find Pydantic model tests
- "Econvention" is the name of an application, not a data element. Use natural names (""Convention", "Activity") for data entities.

LucasZheng42 committed:
chore/test: apply review changes on pydantic models

LucasZheng42 committed:
test: apply reviewer suggestions
feat: add a function which handles date formats
chore: rename some others occurences of oscar/econvention to activity/convention

LucasZheng42 committed:
chore: rename missed occurences of oscar/econvention to activity/convention

LucasZheng42 committed:
chore(tests): test_load_to_oscar.py can now delete its created files to avoid polluting ECONVENTION_TO_OSCAR_OUTPUT_DIR and replace global variables by their fixture counterpart

LucasZheng42 committed:
feat: add additional validation fields in convention_model.py and activity_model.py
test: add corresponding test cases for convention_model.py and activity_model.py
chore: replace direct instantiation with model_validate in transform.py

LucasZheng42 committed:
style: conftest.py and test_load_structure.py applied by blackd

LucasZheng42 committed:
chore: apply reviewer suggestions (rename some functions for more clarity, lower upper_case attributes in convention_model.py and remake the transform logic process)

LucasZheng42 committed:
chore: delete from to transform_econvention_to_oscar.py and rename extract to receive and load to create.
feat: update .env.example to handle some exceptions, transform_econvention_to_oscar.py and create_import_json_to_oscar.py can handle exceptions.
test: add exception testing to verify each task handle exception.

LucasZheng42 committed:
chore: fix comment

LucasZheng42 committed:
fix(deps): replace psycopg2-binary with psycopg2 in requirements.txt because it's better in production
doc: more details about psycopg2 installation

LucasZheng42 committed:
fix(tests): update file path assertion after deps change

LucasZheng42 committed:
chore(activity_model): replace field validators by Milestone and Payment class models
tests: update test_activity_validation.py to assert using model_dump
